File: Editorial/169_Majority_Element/TypeScript/6_Divide_and_Conquer.ts
function majorityElement(nums: number[], lo = 0, hi = null): number {
    let majorityElementRec = (lo: number, hi: number) => {
        if (lo == hi) {
            return nums[lo];
        }

        let mid = Math.floor((hi - lo) / 2) + lo;
        let left = majorityElementRec(lo, mid);
        let right = majorityElementRec(mid + 1, hi);

        if (left == right) {
            return left;
        }

        let leftCount = 0;
        let rightCount = 0;
        for (let i = lo; i <= hi; i++) {
            if (nums[i] == left) {
                leftCount++;
            }
            if (nums[i] == right) {
                rightCount++;
            }
        }

        return leftCount > rightCount ? left : right;
    };

    return majorityElementRec(0, nums.length - 1);
}

File: Editorial/0101-0200/0169_Majority_Element/Go/6_Divide_and_Conquer.go
func majorityElement(nums []int) int {
    var majorityElementRec func(nums []int, lo int, hi int) int
    majorityElementRec = func(nums []int, lo int, hi int) int {
        if lo == hi {
            return nums[lo]
        }

        mid := (hi-lo)/2 + lo
        left := majorityElementRec(nums, lo, mid)
        right := majorityElementRec(nums, mid+1, hi)

        if left == right {
            return left
        }

        leftCount := 0
        rightCount := 0
        for i := lo; i <= hi; i++ {
            if nums[i] == left {
                leftCount++
            }
            if nums[i] == right {
                rightCount++
            }
        }

        if leftCount > rightCount {
            return left
        }

        return right
    }

    return majorityElementRec(nums, 0, len(nums)-1)
}
